Add datastore setup

allthatilk opened the pull request:
Setting up the datastore with Ansible. This is currently tailored to the dev datastore via variable names but this should be an easy thing to update at a later date when we're thinking about production/staging/development stuff.

Some of the roles/task differentiation could use refining but that is a simple thing to do at a later date.

hayfield committed:
Merge pull request #3 from IATI/dev

Update master from dev

dalepotter committed:
Add playbook with initial ansible setup

dalepotter committed:
Add install requirements and postgres setup commands to Datastore setup

dalepotter committed:
Add installation of dependency

dalepotter committed:
Update server IP address

dalepotter committed:
Reload bash profile to activate environment variable

dalepotter committed:
Run command to create datastore database tables

allthatilk committed:
Fix errors preventing table creation

Debugged to find setup.py in Datastore repo needs updating for new verison of six
Fixed permissions error to allow table creation
Added task to install python-minimal to server on first run
Fixed environment variable problem by adding at task level
Prevented facts_gathering to allow python-minimal to install before other tasks run

allthatilk committed:
Add iati crawl update command

Add new command
Refactor environment variable to play level for future database commands

allthatilk committed:
Add cron jobs and apache2 config things

Set up crawler to run on the half hour every 5 hours
Set up worker to download and index data files
Add apache2 requirements for deployment: config file and apt install requirements and dependencies

allthatilk committed:
WIP Fixed ansible error but site not working

Add default conf files but not doing anything useful yet
Add datastore user grou but again, not doing anything useful yet
Enable mod_headers to fix apache2 error in ansible
Add check for server reboot
Add handler for apache2 restart
Remove commented out code in apache2.conf

allthatilk committed:
latest changes

dalepotter committed:
Add task to grant appropriate database permissions

This ensures that the datastore user can access tables in the `iati-datastore` database.

dalepotter committed:
Merge pull request #8 from IATI/datastore-setup-fixes

Add task to grant appropriate database permissions

allthatilk committed:
Add missing env variables for cron jobs

allthatilk committed:
Add sequence permissions and remove unnecessary task

allthatilk committed:
Remove unnecessary tasks

allthatilk committed:
Begin role separation

Split roles into apache2, postgres, python2, and datastore, may need more refining later.
Add group_vars all to set default python interpreter to python3
Set python interpreter to python2 if python2 role included in play

allthatilk committed:
Refactor out most unique variables

Extract variables from apache2 role, and some from datastore role

allthatilk committed:
Refactor code to emulate Ansible best practice (mostly)

Separate out more roles from current tasks
Extract variables where appropriate (mostly)
Move roles to most relevant locations
Move apache2 file to apache role files folder
Update all tasks files references to roles and other tasks
Update datastore.yml to reflect new structure
Update README to fix existing errors
Update README to include Ansible 2.4 requirement and info
Update README to include info about Datastore OS compatibility

hayfield commented on README.md:
```diff
@@ -10,33 +10,36 @@ Based on guide at http://ryaneschinger.com/blog/ansible-quick-start/
-ansible all --inventory-file=inventory.ini --module-name ping -u root
-ansible all -i inventory.ini --m ping -u root
+ansible all --inventory-file=servers.ini --module-name ping -u root
+ansible all -i servers.ini -m ping -u root
```
Having both of these looks unnecessary

hayfield commented on README.md:
```diff
@@ -10,33 +10,36 @@ Based on guide at http://ryaneschinger.com/blog/ansible-quick-start/
-ansible all -i inventory.ini -m command -u root --args "uptime"
-ansible all -i inventory.ini -m command -u root -a "uptime"
+ansible all -i servers.ini -m command -u root --args "uptime"
+ansible all -i servers.ini -m command -u root -a "uptime"
```
Having both of these looks unnecessary

hayfield commented on roles/dev_datastore/datastore_setup/vars/main.yml:
```diff
@@ -0,0 +1,18 @@
+
+user_home_dir: /home/datastore/
+
+repo_name: IATI-Datastore/
```
If a repo name, shouldn't include the `/` (that's a folder path)

hayfield commented on roles/dev_datastore/datastore_setup/vars/main.yml:
```diff
@@ -0,0 +1,18 @@
+  - libpq-dev
+  - libxml2-dev
+  - libxslt-dev
+  - libevent-dev
```
Out of alphabetical order

hayfield commented on roles/common/postgres/tasks/main.yml:
```diff
@@ -0,0 +1,31 @@
+
+- name: Create postgres role for custom user
+  become: true
+  become_user: postgres
```
The required `postgres` user is not created within the `postgres` role (so the role cannot be used on its own).

allthatilk replied:
I don't believe `postgres` is a user that needs creating. I think it's set up automatically somewhere so will add a comment to clarify this.

hayfield commented on roles/common/postgres/tasks/main.yml:
```diff
@@ -0,0 +1,31 @@
+    state: present
+    update_cache: yes
+
+- name: Install psycopg2
```
This looks like a very specific thing. Is it an ansible requirement? If so, a comment would be useful.

hayfield commented on roles/common/postgres/tasks/main.yml:
```diff
@@ -0,0 +1,31 @@
+- name: Set environment variable to link to the postgres database
+  lineinfile:
+    dest: /etc/environment
+    line: "export DATABASE_URL=postgres:///{{ custom_database }}"
```
The role is fairly generic, though use of `DATABASE_URL` means there can only be a single DB on a server like this. Suggest turning this into a variable.

allthatilk replied:
Leaving this suggestion for a later iteration due to complications with ansible env vars

hayfield commented on roles/dev_datastore/apache2/handlers/main.yml:
```diff
@@ -0,0 +1,5 @@
+
```
Missing `---`

hayfield commented on roles/common/install_dependencies/tasks/main.yml:
```diff
@@ -0,0 +1,8 @@
+    state: present
+    update_cache: yes
+  with_items:
+     "{{ packages }}"
```
The variable name `packages` isn't entirely clear what it's used for. Suggest something like `required_packages` or `package_dependencies` or a better idea than those suggestions to clarify what the variable is for.

allthatilk replied:
`required_packages`

allthatilk replied:
done

hayfield commented on roles/common/postgres/tasks/main.yml:
```diff
@@ -0,0 +1,31 @@
+  become: true
+  become_user: postgres
+  postgresql_user:
+    name: "{{ custom_user }}"
```
It may be desired to have multiple users doing different tasks.

Suggest changing the variable name to indicate that it's to be the owner of a postgres database.

allthatilk replied:
done

hayfield commented on roles/common/postgres/tasks/main.yml:
```diff
@@ -0,0 +1,31 @@
+  become: true
+  become_user: postgres
+  postgresql_db:
+    name: "{{ custom_database }}"
```
Suggest adding a `_name` suffix to the variable to better clarify that it's the name of the database.

allthatilk replied:
done

hayfield commented on roles/common/python2/tasks/main.yml:
```diff
@@ -0,0 +1,10 @@
+- name: Install python 2
+  raw: test -e /usr/bin/python || (apt -y update && apt install -y python-minimal)
+
+- name: Install pip
```
The `install_dependencies` role should be usable here.

allthatilk replied:
done

hayfield commented on roles/common/restart_server/tasks/main.yml:
```diff
@@ -0,0 +1,9 @@
+---
+- name: Restart server # create role for reboot
```
Is the comment needed any more?

hayfield commented on roles/common/restart_server/tasks/main.yml:
```diff
@@ -0,0 +1,9 @@
+  async: 1
+  poll: 0
+
+- name: Wait 30 seconds for server to come back
```
The description says 30s, while the action says 20.
Suggest updating the description to remove the number.

hayfield commented on roles/dev_datastore/apache2/tasks/main.yml:
```diff
@@ -0,0 +1,23 @@
+- include_role:
+    name: common/install_dependencies
+
+- name: Enable mod_headers
```
Different installations of `apache` will require different modules. Suggest extracting the list of required modules into a variable and using `{{ items }}`.

hayfield commented on roles/dev_datastore/datastore_setup/vars/main.yml:
```diff
@@ -0,0 +1,18 @@
+
+github_repo: https://github.com/IATI/IATI-Datastore.git
+
+user_home_dir: /home/datastore/
```
The username part of this looks like something that should be able to come from a variable containing the name of the user.

hayfield commented on roles/dev_datastore/datastore_setup/vars/main.yml:
```diff
@@ -0,0 +1,18 @@
+
+repo_name: IATI-Datastore/
+
+wsgi_file: datastore.wsgi
```
Suggest adding `_name` suffix.

allthatilk committed:
Update README by removing duplicate code examples

allthatilk committed:
Change repo_name to repo_dir for accuracy

allthatilk committed:
Put packages in alphabetical order

allthatilk committed:
Add comment about postgres user

allthatilk committed:
Add comment about psychopg2

allthatilk committed:
Remove outdated comment and update task name

allthatilk committed:
Condense apache module install to one task and extract variables

allthatilk committed:
Rename wsgi_file to wsgi_file_name, swap 'datastore' in user_home_dir variable for extracted variable

allthatilk commented:
@hayfield ready for re-review

hayfield commented on datastore.yml:
```diff
@@ -0,0 +1,16 @@
+- hosts: datastore-dev
+  remote_user: root
+  gather_facts: no
+  roles:
```
Does this want the `config` role to be run?

allthatilk replied:
Nope.

allthatilk replied:
nope

hayfield commented on datastore.yml:
```diff
@@ -0,0 +1,16 @@
+---
+- hosts: datastore-dev
```
`dev` or `staging`?

Should ensure names are consistent.

allthatilk replied:
dev

allthatilk replied:
done

hayfield commented on datastore.yml:
```diff
@@ -0,0 +1,16 @@
+    - common/python2
+    - common/postgres
+    - common/create_user
+    - dev_datastore/datastore_setup
```
Is this for all `datastore` instances rather than just a `dev datastore`?

allthatilk replied:
datastore

hayfield commented on datastore.yml:
```diff
@@ -0,0 +1,16 @@
+    - dev_datastore/datastore_setup
+
+  environment:
+    DATABASE_URL: postgres:///iati-datastore
```
Can the `custom_database` variable be used to specify the database name?

allthatilk replied:
yes

allthatilk committed:
Add db_user, fix wsgi_file_name find-replace error

allthatilk committed:
Rename playbook to be specific to dev datastore, genericize datastore role

allthatilk committed:
Extract some of DATABASE_URL variable

allthatilk committed:
Change dependency variables

Use common/install_dependecies where possible
Rename packages to required_packages
Create vars folder for postgres variables
Add dasshes to start of handler file

hayfield commented on datastore_dev.yml:
```diff
@@ -0,0 +1,19 @@
+    - datastore/datastore_setup
+
+  environment:
+    DATABASE_URL: "{{ db_env_var }}"
```
Variable name could be clarified to better demonstrate what it is to be used for than just being an environment variable for a database.

allthatilk replied:
This needs some thought.

hayfield commented on roles/datastore/datastore_setup/tasks/setup_datastore.yml:
```diff
@@ -0,0 +1,20 @@
+---
+- name: Create the iati-datastore tables
```
Would look to split this into two `yml` files - one to create the various tables in the database, the other to add things to cron.
(from a maintenance point of view, it isn't immediately clear that `setup_datastore.yml` is the place to go to update the crontab)

allthatilk replied:
done

hayfield commented on roles/datastore/datastore_setup/tasks/update_crawler.yml:
```diff
@@ -0,0 +1,3 @@
+---
+- name: "Update source data with crawler"
```
This command is run by `cron`, so isn't totally clear why it's also a separate task.

Suggest adding a comment to explain why it has to be here too.

hayfield commented on roles/common/clone_repo/tasks/main.yml:
```diff
@@ -0,0 +1,6 @@
+---
+- name: Clone the repository
+  git:
+    repo: "{{ github_repo }}"
```
Suggest changing the variable name to better clarify that it's a cloneable `.git` address (and that the role could take a repo from bitbucket or something else that isn't github).

`git_clone_url` or `git_clone_address` or something along those lines maybe?

allthatilk replied:
`git_clone_address`

hayfield commented on test:
```diff
@@ -0,0 +1,3 @@
+# no longer using update -t 2 because of https://github.com/IATI/iati-datastore/issues/194
```
Is this file still used for anything, or can it be removed?

allthatilk replied:
no idea. should not be there.

hayfield reviewed (changes requested):
A small number of lil things :)

allthatilk committed:
Add comment to clarify why duplicate crawler command is needed.

allthatilk committed:
Make database variable name more descriptive

allthatilk committed:
Separate responsibility for cron jobs

allthatilk committed:
Make git_repo name more descriptive

allthatilk committed:
Remove unnecessary test file

hayfield commented on roles/datastore/apache2/tasks/main.yml:
```diff
@@ -0,0 +1,21 @@
+---
+# currently residing in dev-datastore rather than common as only needed by datastore. Can be extracted at a later date if necessary.
```
Comment refers to old role name.

allthatilk replied:
Ah, will fix.

hayfield reviewed (approved):
tiny thing :peanuts:

allthatilk committed:
Update comment to reflect role name change
